src/Lecture2.hs

```diff
@@ -158,17 +175,55 @@ data Knight = Knight
     { knightHealth    :: Int
     , knightAttack    :: Int
     , knightEndurance :: Int
-    }
+    } deriving (Show)
+
+data Chest a = Chest {
+  gold :: Int,
+  treasure :: (Maybe a) 
+} deriving (Show)
+
+data DragonType = Red|Black|Green deriving (Show)
+data Dragon a = Dragon {
+    dragonHealth :: Int,
+    dragonAttack :: Int,  
+    dragonTreasure :: Chest a,
+    dragonColor:: DragonType
+} deriving (Show)
+
+type Reward a = (Int, Int, Maybe a)
```

Comment:
This could be a record with some newtypes for more type-safety and readability 🙂